File: imu_calibrator.rb
#!/bin/env ruby

require 'matrix'

module Profiler    
  require 'ruby-prof'

  def run_time
    start = Time.now
    yield
    Time.now - start
  end
  
  def profile(&block)
    result = RubyProf.profile &block
    print_profile_result result
    result
  end
  
  def print_profile_result(result)
    File.open("profile.html", "w") do |file|
      RubyProf::GraphHtmlPrinter.new(result).print(file)
    end
  end
end
  
include Profiler

module Enumerable
  def percentile_to_value(*percentiles)
    sorted_values = sort
    percentiles.collect do |p|
      index = (p / 100.0 * (sorted_values.size-1)).round
      sorted_values[index]
    end
  end
  
  def average
    inject(:+) / size.to_f
  end
  
  def variance
    m = average
    sum = inject(0){|accum, i| accum + (i-m)**2 }
    sum/(size - 1).to_f
  end

  def std_deviation
    Math.sqrt variance
  end
  
  def regular_sample(num)
    each_slice(size/num).collect(&:first)
  end
  
end

class Calibration

  attr_reader :values, :raw_readings
  
  def initialize(values, raw_readings=nil)
    @values = values
    @values.freeze
    @raw_readings = raw_readings
  end
  
  def switch_readings(readings)
    self.class.new values, readings
  end
  
  def increment(value_id, change)
    new_values = values.dup
    new_values[value_id] += change
    self.class.new new_values, raw_readings
  end

  
  def score
    @score ||= -scaled_magnitudes.collect{ |m| (m - 1.0)**2 }.average
  end
  
  def scaled_magnitudes
    @scaled_magnitudes ||= scaled_readings.collect(&:magnitude)
  end
  
  def scaled_readings
    @scaled_readings ||= raw_readings.collect { |r| scale r }
  end
  
  def scale(raw_reading)
    Vector[(raw_reading[0] - values[0])/(values[1] - values[0]).to_f * 2 - 1,
      (raw_reading[1] - values[2])/(values[3] - values[2]).to_f * 2 - 1,
      (raw_reading[2] - values[4])/(values[5] - values[4]).to_f * 2 - 1]
  end

  def info_string
    "%-32s %7.4f %7.4f %7.4f" % [
      to_s,
      scaled_magnitudes.average,
      scaled_magnitudes.std_deviation,
      score
    ]
  end
  
  def to_s
    "%d %d %d %d %d %d" % values
  end
end

class ImuCalibrator
  Axes = (0..2)

  def run(file=$stdin)
    raw_readings = read_vectors(file).freeze
    raw_readings_sample = raw_readings.regular_sample(300).freeze
    cal1 = guess(raw_readings)
    cal2 = tune(cal1, raw_readings_sample)
    cal3 = tune(cal2, raw_readings)
    puts cal3
  end
  
  def read_vectors(file)
    vectors = file.each_line.collect do |line|
      coords = line.split(/,?\s+/).reject(&:empty?).first(3).collect(&:to_i)
      Vector[*coords]
    end
    
    vectors.uniq!  # save processing time
    
    vectors
  end
  
  def guess(readings)
    guess = Axes.flat_map do |axis|
      values = readings.collect { |v| v[axis] }
      values.percentile_to_value(1, 99)
    end
    Calibration.new guess
  end
  
  def tune(cal, readings)
    cal = cal.switch_readings(readings)
    $stderr.puts cal.info_string
    while true
      last_cal = cal
    
      cal.values.each_index do |value_id|
        cal = try_dir(cal, value_id, 1) || try_dir(cal, value_id, -1) || cal
      end
      $stderr.puts cal.info_string
      
      return cal if last_cal == cal
    end
  end

  def try_dir(cal, value_id, dir)
    improved_cal = nil
    while true
      new_cal = cal.increment(value_id, dir)
      return improved_cal unless new_cal.score > cal.score
      improved_cal = cal = new_cal
    end
  end
  
end

profile do
  ImuCalibrator.new.run
end

File: imu_calibrator.py
from __future__ import print_function
import sys
import math

# TODO: this shouldn't work, right???  Becuase cache will not be instance-specific
# try: http://code.activestate.com/recipes/577452-a-memoize-decorator-for-instance-methods/
def memoize(f):  # change this to assume no args are provided and simplify it
  cache = {}
  def memf(*x):
    if x not in cache:
      cache[x] = f(*x)
    return cache[x]
  return memf

def average(list):
  return sum(list)/len(list)

def percentile_to_value(list, *percentiles):
  list = sorted(list)
  return [list[int(p / 100.0 * (len(list)-1))] for p in percentiles]

def variance(list):
  m = average(list)
  return sum([(i-m)**2 for i in list]) / float(len(list))

def std_deviation(list):
  return math.sqrt(variance(list))

class Vector(object):
  def __init__(self, x, y, z):
    self.x, self.y, self.z = x, y, z
  def __str__(self):
    return "(" + str(self.x) + ", " + str(self.y) + ", " + str(self.z) + ")"      

  def __getitem__(self, key):
    if key == 0:
      return self.x
    elif key == 1:
      return self.y
    elif key == 2:
      return self.z
    else:
      raise Exception("Invalid key")
      
  def magnitude(self):
    return math.sqrt(self.x**2 + self.y**2 + self.z**2)
    
Axes = range(3)

class Calibration:
  def __init__(self, values, raw_readings=None):
    self.values = values
    self.raw_readings = raw_readings

  def scale(self, raw_reading):
    return Vector((raw_reading[0] - self.values[0])/float(self.values[1] - self.values[0]) * 2 - 1,
      (raw_reading[1] - self.values[2])/float(self.values[3] - self.values[2]) * 2 - 1,
      (raw_reading[2] - self.values[4])/float(self.values[5] - self.values[4]) * 2 - 1)

  def __str__(self):
    return "%d %d %d %d %d %d" % tuple(self.values)

  def info_string(self):
    return "%-32s %7.4f %7.4f %7.4f" % (
      str(self),
      average(self.scaled_magnitudes()),
      std_deviation(self.scaled_magnitudes()),
      self.score()
    )
    
  def switch_readings(self, raw_readings):
    return Calibration(self.values, raw_readings)
  
  @memoize
  def scaled_magnitudes(self):
    return [s.magnitude() for s in self.scaled_readings()]
    
  @memoize
  def scaled_readings(self):
    return [self.scale(r) for r in self.raw_readings]
  
  @memoize
  def score(self):
    return -average([(m - 1.0)**2 for m in self.scaled_magnitudes()])
  
  def increment(self, value_id, dir):
    values = list(self.values)
    values[value_id] += dir
    return Calibration(values, self.raw_readings)
    
  
def run(file=sys.stdin):
  raw_readings = read_vectors(file)
  raw_readings_sample = raw_readings[0::(len(raw_readings)/300)]
  cal1 = guess(raw_readings)
  cal2 = tune(cal1, raw_readings_sample)
  cal3 = tune(cal2, raw_readings)  
  print(cal3)

def read_vectors(file):
  return [Vector(*[int(s) for s in line.split()[0:3]]) for line in file]

def guess(readings):
  guess = []
  for axis in Axes:
    values = [v[axis] for v in readings]
    guess.extend(percentile_to_value(values, 1, 99))
  return Calibration(guess)

def try_dir(cal, value_id, dir):
  improved_cal = None
  while True:
    new_cal = cal.increment(value_id, dir)
    if not new_cal.score() > cal.score():
      return improved_cal
    improved_cal = cal = new_cal
  return None
  
def tune(cal, readings):
  cal = cal.switch_readings(readings)
  print(cal.info_string(), file=sys.stderr)
  while True:
    last_cal = cal
    
    for value_id in range(len(cal.values)):
      cal = try_dir(cal, value_id, 1) or try_dir(cal, value_id, -1) or cal
    print(cal.info_string(), file=sys.stderr)

    if last_cal == cal:
      return cal 
    
  
  
if __name__=='__main__':
  run()